src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/FileServerHandler.java

```diff
@@ -327,14 +327,34 @@ private static String getLastModified(Path path) throws IOException {
         return fileTime.toInstant().atZone(ZoneId.of("GMT")).format(HTTP_DATE_FORMATTER);
     }
 
+    // Checks if this path's or any of its subpaths' files are hidden or symlinks.
     private static boolean isHiddenOrSymLink(Path path) {
         try {
-            return Files.isHidden(path) || Files.isSymbolicLink(path);
+            int segmentCount = path.getNameCount();
+            for (int i = 1; i <= segmentCount; i++) {
+                Path p = path.getRoot().resolve(path.subpath(0, i));
+                if (Files.isHidden(p) || Files.isSymbolicLink(p)) {
+                    return true;
+                }
+            }
+            return false;
```

Comment:
This will start checking from the root of the file system. I believe we want to start checking from the root of the FileServerHandler, root excluded.

Reply:
Maybe these checks should be made in `mapToPath` instead since you already walk the path there - and IIRC returning null from `mapToPath` will cause HTTP 404.

Reply:
Agreed. I refactored the handler to check `!isReadable`, `isHidden` and `isSymbolicLink` for each path segment from the root, excluding the root itself. If any of these conditions is met, a 404 response is sent. A generic 404 response for hides potentially revealing information. The checks are repeated in handle() on line 375, let's keep them in there for reassurance.